### ISSUE TITLE
`withUnpublishedPreview` always shows 404 page

### ISSUE THREAD

**Issue by seaneking** [angeloashmore/gatsby-source-prismic#246]
Not sure if I'm just missing something very obvious, but I can't get the `withUnpublishedPreview` hook to work. No matter what I do it just loads the 404 page, albeit at the correct preview URL.

At first I thought it was because my 404 page is also using Prismic data and the preview data wasn't resolving properly, but even stripping all that out and just using static fallback markup it doesn't work.

Here's my 404 page, edited for succinctness:

```js
import { graphql } from 'gatsby';
import { withUnpublishedPreview } from 'gatsby-source-prismic';
import React from 'react';
import ProjectTemplate from '../templates/project';

export default withUnpublishedPreview(NotfoundPage as any, {
  templateMap: {
    project: ProjectTemplate
  }
});

function NotfoundPage({ data: query }: { data: GatsbyTypes.NotfoundQuery }) {
  const { data } = query.prismicNotfound!;

  return (
    <>
      // 404 page markup here
    </>
  );
}

export const query = // query here
`;
```

And my preview page

```js
import {
  withPreviewResolver,
  WithPreviewResolverProps
} from 'gatsby-source-prismic';
import React, { useState } from 'react';
import { resolveDocument } from '../lib/resolve';

declare global {
  interface Window {
    __PRISMIC_PREVIEW_DATA__: any;
  }
}

export default withPreviewResolver(PreviewPage as any, {
  repositoryName: 'karimjoreige',
  linkResolver: (): any => resolveDocument
});
function PreviewPage({ isPreview }: WithPreviewResolverProps) {
  const [message, setMessage] = useState('Loading preview...');
  if (isPreview === false) {
    setMessage(`Can't find preview`);
  }

  return (
    <div style={{ alignSelf: 'center', justifySelf: 'center' }}>{message}</div>
  );
}
```

**Comment by angeloashmore** [comment]
Hi @seaneking, you should be able to use Prismic data on the 404 page.

Couple of questions, some which I think you already answered:

- Is the redirected URL correct? I.e. the URL is what the unpublished document would have as if it were published?
- Is the document type `project`?
- Any console errors or warnings?
- If everything sounds correct, could you use the React DevTools to inspect the `<PreviewStoreProvider>` to see if the store contains preview data for the previewed document's URL? You should see a state object with a key named `pages`.

**Comment by seaneking** [comment]
> Is the redirected URL correct? I.e. the URL is what the unpublished document would have as if it were published?

Yep 

> Is the document type project?

Yep

> Any console errors or warnings?

Yep there's one warning that I missed before (other than the 404s for the page and related `page-data` requests):

```
An unpublished preview was detected, but a template component could not be found for a custom type of "prismicNotfound"
``` 

`prismicNotFound` is my 404 page's Prismic document, but I'm trying to preview a `project` document. As an aside, I assumed from the example in the docs + the format of the `schemas` map in `gatsby-config` that I'd map templates to basic prismic types, like `project` rather than `prismicProject`. This error suggests I should be using the GraphQL node types instead? Either way tried using `prismicProject` in my `templateMap` and it doesn't work.

> If everything sounds correct, could you use the React DevTools to inspect the <PreviewStoreProvider> to see if the store contains preview data for the previewed document's URL? You should see a state object with a key named pages.

Yep, looks like it has data, screenshot below:

![Screen Shot 2020-06-29 at 7 17 53 PM](https://user-images.githubusercontent.com/6374876/85984446-503a7080-ba3d-11ea-8970-5555856c53de.png)

The Prismic Toolbar is also showing the correct document

![Screen Shot 2020-06-29 at 7 29 24 PM](https://user-images.githubusercontent.com/6374876/85985474-e4f19e00-ba3e-11ea-99ff-b60a8580717a.png)

**Comment by seaneking** [comment]
Update — it only seems to have data, show the toolbar, and have that weird warning in a development environment. In production it just 404s silently, but at the correct URL.

**Comment by seaneking** [comment]
Anything else I can do to debug this? It’s got me stumped

**Comment by angeloashmore** [comment]
@seaneking Sorry, haven't had a chance to look into this yet. I should have some time over the weekend.

If you have time to debug, I'd start with this line here:
https://github.com/angeloashmore/gatsby-source-prismic/blob/1fb9d74fa9145d19e8f5da513eab48b38e05a1d1/src/withUnpublishedPreview.tsx#L21-L26

Looking at it now, that `console.warn` should actually be in an `else` after the `if (TemplateComp)`; it's probably warning even for working previews.

**Comment by angeloashmore** [comment]
I did some experimenting and came to the conclusion that the issue is more nuanced than I described above.

The problem is two-fold:

## `mergePrismicPreviewData`

`mergePrismicPreviewData` assumes it can check for the existence of the same custom type in the static data as the preview data. For example, if you are previewing a document with an API ID of `project`, `mergePrismicPreviewData` will look for `prismicProject` at the root of `props.data` (i.e. the static Gatsby data).
   - **If present**: replace the static version with the preview version.
   - **If not present**: traverse the data and find an object containing a matching `prismicId` field (the document ID from Prismic's API). If it finds a match, replace that object with the preview data. This is necessary to replace/preview documents nested within other documents/queries.

### Proposed fix

One possible solution is to only do the "if not present" flow. In other words, preview data will only replace static data where it can find a matching `prismicId` field. The current system checks for a matching custom type field (e.g. `prismicProject`), but this is too imprecise to replace nested documents. This would mean requiring developers to add a special field to their queries if they want to enable previews for the document.

<details>
<summary><strong>Open example query</strong></summary>

```graphql
query PageTemplate($uid: String!) {
  prismicPage(uid: { eq: $uid }) {
    # Required to enable previews for this document
    prismicId
    data {
      title {
        text
      }
      image {
        url
      }
    }
  }
}
```
</details>

Requiring `prismicId` seems a bit arbitrary from an end-user perspective, so it could be aliased with a more descriptive name:

<details>
<summary><strong>Open example query</strong></summary>

```graphql
query PageTemplate($uid: String!) {
  prismicPage(uid: { eq: $uid }) {
    # Required to enable previews for this document
    _previewable
    data {
      title {
        text
      }
      image {
        url
      }
    }
  }
}
```
</details>

With a custom directive/extension, I think it's possible to also move the preview marker out of the fields:

<details>
<summary><strong>Open example query</strong></summary>

```graphql
query PageTemplate($uid: String!) {
  prismicPage(uid: { eq: $uid }) @previewable {
    data {
      title {
        text
      }
      image {
        url
      }
    }
  }
}
```
</details>

This would only need to be added at the first level of document fields. Here's an example of what a more complex query could look like:

<details>
<summary><strong>Open example query</strong></summary>

```graphql
query PageTemplate($uid: String!) {
  prismicPage(uid: { eq: $uid }) {
    # Required to enable previews for this document
    _previewable
    data {
      title {
        text
      }
      image {
        url
      }
      related_blog_post {
        document {
          ... on PrismicBlogPost {
            data {
              # Required to enable previews for this document
              _previewable
              title {
                text
              }
              excerpt
            }
          }
        }
      }
    }
  }
}
```
</details>

So the question is, **would you be okay with needing to add `_previewable` or `@previewable` to your query in order to enable previews?** This would allow for a stronger preview system.

## `withUnpublishedPreview`

`withUnpublishedPreview` assumes the document to preview is at the root of the template's query. For example, if you preview an unpublished `project` document, it assumes the template you provide expects `prismicProject` to be queried at `props.data.prismicProject`. `withUnpublishedPrevew` has no way of knowing where else to place that document in `props.data`.

The data will run through `withPreview`, which runs through `mergePrisimicPreviewData` using the logic explained above. In @seaneking's case, this means the merge function *will not* find a key named `prismicProject` since it only contains `prismicNotFound`. It will then look for an object containing a matching `prismicId`, which it also does not find. Thus, it does not merge the preview data.

In the sites I've worked on where we query a document on the 404 page and perform previews, everything used the same custom type so it was able to find a matching key.

## Proposed solution

To support structures like @seaneking's, I think we need to add logic to `withUnpublishedPreview` to always add the previewed data to the root of the document. If the key already exists, it can replace it. If it doesn't exist, it is added.

---

@seaneking Would you be able to provide some thoughts on the proposals? Both should be doable without breaking changes, but with warnings that there will be removed functionality in the next major version.

**Comment by angeloashmore** [comment]
PR with these changes is here: https://github.com/angeloashmore/gatsby-source-prismic/pull/247

**Comment by seaneking** [comment]
Wow thanks for the deep dive @angeloashmore! Looks fantastic. 

I think patching `withUnpublishedPreview` to always add previewed data, and replacing existing if it needs to, is a no brainer. 

Adding a `@previewable` directive (which would be my preference compared to a `_previewable` key, since that feels like a mixing of userland data and source logic. I think requiring `prismicId` is more transparent req if not using the custom directive) is a bit annoying, but if it makes for a stronger preview system then I think it's worth the extra keystrokes. Prismic previews are such a huge pain in the other source available, so anything to make them more solid is a win in my books. Especially if it's a non-breaking change, you could always add it and throw it behind an experimental flag until the next major release to test the waters?

**Comment by angeloashmore** [comment]
Going to close this out since I think this was resolved in #247 with the `v3.2.0` release.

Let me know if this is still an issue though. 👍
